tempesta_fw/lb/tfw_lb_http.c

```diff
@@ -0,0 +1,1091 @@
+ * The slot may be accessed via TFW_LB_GROUP_RR_CTR(TfwLbGroup->rr_idx).
+ *
+ * XXX: non-shared round-robin counters allow to avoid atomic instructions
+ * and prevent cache line bouncing.
+ */
+static DEFINE_PER_CPU(u8[TFW_LB_GROUP_MAX], tfw_lb_group_rr_ctrs);
+
+#define TFW_LB_GROUP_RR_CTR(group) \
+	(__get_cpu_var(tfw_lb_group_rr_ctrs)[(group)->rr_idx])
+
+/**
+ * The number entries in the tfw_lb_groups list.
+ * Helps to protect tfw_lb_group_rr_ctrs from overflow.
+ */
+static int tfw_lb_groups_n;
+
+/*
+ * ------------------------------------------------------------------------
+ *	tfw_lb_http_send_msg() and its scheduling helpers
+ * ------------------------------------------------------------------------
+ *
+ * This is the actual load balancing implementation.
+ * The code decides how to distribute HTTP requests across backend servers.
+ */
+
+static struct sock *
+tfw_lb_sched_rr_failover(TfwLbGroup *group)
+{
+	int sks_n, ctr, ctr_end;
+	struct sock *sk, **sks;
+
+	sks = group->sks;
+	sks_n = group->sks_n;
+	ctr = TFW_LB_GROUP_RR_CTR(group);
+	ctr_end = ctr + sks_n;
+
+	while (likely(++ctr < ctr_end)) {
+		sk = sks[ctr % sks_n];
+		if (likely(sk->sk_state == TCP_ESTABLISHED)) {
+			TFW_LB_GROUP_RR_CTR(group) = ctr;
+			return sk;
+		}
+	}
+
+	TFW_ERR("can't find any online backend server\n");
+	return NULL;
+}
+
+/**
+ * XXX: the whole tfw_sched_rr is collapsed into this single function.
+ */
+static struct sock *
+tfw_lb_sched_rr(TfwLbGroup *group)
+{
+	struct sock *sk;
+
+	sk = group->sks[++TFW_LB_GROUP_RR_CTR(group) % group->sks_n];
+	if (likely(sk->sk_state == TCP_ESTABLISHED))
+		return sk;
+
+	return tfw_lb_sched_rr_failover(group);
+}
+
+/**
+ * XXX: the whole tfw_sched_hash is collapsed into this single function.
+ * Rendezvous hashing is included as a failovering strategy.
+ *
+ * XXX: the linear search is slow but it is required for failover switching.
+ * Perhaps we should do binary search first, and then fall back to linear if
+ * the chosen socket is dead.
+ *
+ * XXX: schedulers will be separated into modules later. Now the code is here
+ * because it determines how to build everything else around it.
+ */
+static struct sock *
+tfw_lb_sched_hash(TfwLbGroup *group, TfwHttpReq *req)
+{
+	struct sock **sk_pos, *sk, *best_sk;
+	u32 *hash_pos, hash, best_hash, req_hash;
+
+	best_sk = NULL;
+	best_hash = 0;
+	req_hash = tfw_http_req_key_calc(req);
+
+	sk_pos = group->sks;
+	hash_pos = group->sk_hashes;
+	do {
+		sk = *sk_pos;
+		hash = *hash_pos;
+
+		if ((req_hash ^ hash) > (req_hash ^ best_hash) &&
+		    likely(sk->sk_state == TCP_ESTABLISHED)) {
+			best_sk = sk;
+			best_hash = hash;
+		}
+	} while (*(++sk_pos) && *(++hash_pos));
+
+	return best_sk;
+}
+
+static struct sock *
+tfw_lb_sched(TfwLbGroup *group, TfwHttpReq *req)
+{
+	/* Should be done via a function pointer, but KISS for now. */
+	return (group->sched == TFW_LB_SCHED_RR)
+	       ? tfw_lb_sched_rr(group)
+	       : tfw_lb_sched_hash(group, req);
```

Comment:
Good improvement - it's much better to have single powerful scheduler rather than 3 simple schedulers. However, it's still better to call the scheduling functions by callbacks, don't postpone it.